### PR TITLE
Amoc coordinator

### PR DESCRIPTION
**Pull request by DenysGonchar**
amoc_coordinator allows to group users in a simple manner

### CODE REVIEW

File: src/amoc_coordinator.erl [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Whenever a new event handler is added to an event manager,
+%% this function is called to initialize the event handler.
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec(init({timeout, atom(), coordination_timeout_in_sec()} | normalized_coordination_item()) ->
```
Comment (michalwski) [inline src/amoc_coordinator.erl]: What is the benefit of using this `spec` syntax over the other syntax used in previous lines? Is it possible to be consistent?

File: src/amoc_coordinator.erl [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Whenever an event manager receives an event sent using
+%% gen_event:notify/2 or gen_event:sync_notify/2, this function is
+%% called for each installed event handler to handle the event.
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec(handle_event(Event :: term(), state()) -> {ok, state()}).
```
Comment (michalwski) [inline src/amoc_coordinator.erl]: Another spec syntax inconsistency.

File: src/amoc_coordinator.erl [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Whenever an event manager receives a request sent using
+%% gen_event:call/3,4, this function is called for the specified
+%% event handler to handle the request.
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec(handle_call(Request :: term(), state()) -> {ok, {error, not_implemented}, state()}).
```
Comment (michalwski) [inline src/amoc_coordinator.erl]: Another spec syntax inconsistency.

File: src/amoc_coordinator.erl [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+%% @private
+%% @doc
+%% Whenever an event handler is deleted from an event manager, this
+%% function is called. It should be the opposite of Module:init/1 and
+%% do any necessary cleaning up.
+%%
+%% @spec terminate(Reason, State) -> void()
+%% @end
+%%--------------------------------------------------------------------
+-spec(terminate(any(), state()) -> ok).
```
Comment (michalwski) [inline src/amoc_coordinator.erl]: Another spec syntax inconsistency.

File: src/amoc_coordinator_worker.erl [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+                n = 0 :: non_neg_integer(),
+                actions = [] :: [action()],
+                collect_data = true :: boolean(),
+                accumulator = [] :: [data()]}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+-spec(start(amoc_coordinator:normalized_coordination_item()) -> {ok, Pid :: pid()}).
```
Comment (michalwski) [inline src/amoc_coordinator_worker.erl]: Is it possible to use one spec syntax? At least across single PR.

File: src/amoc_coordinator_worker.erl [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+start(CoordinationItem) -> gen_server:start(?MODULE, CoordinationItem, []).
```
Comment (michalwski) [inline src/amoc_coordinator_worker.erl]: Could you explain why we use `start` and not `start_link` here?

**Comment by michalwski** [comment]
Oh, one more thing. Is this meant to be used in single AMOC node or in a cluster?

File: src/amoc_coordinator.erl [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            %%    handle_event(Event, State) for each installed event
+            %%    handler, in the same order as they were added.
+            %%
+            %% in reality the order is reversed, the last added handler
+            %% is executed at first. so to ensure that all the items in
+            %% the plan with NoOfUsers =:= all are executed in the very
+            %% end, we need to add them first.
+            AllItemsHandlers = lists:reverse([Item || {all, _} = Item <- Plan]),
+            [gen_event:add_handler(Name, ?MODULE, Item) || Item <- AllItemsHandlers],
+            [gen_event:add_handler(Name, ?MODULE, Item) || {N, _} = Item <- Plan, is_integer(N)],
```
[the remark this answers is not in the document]
Comment (asas12) [reply src/amoc_coordinator.erl]: If I am not mistaken they won't be, as they are denoted with `all` as the first element of the tuple. They will be added from `AllItemsHandlers` and not added in the next line, as they are filtered out based on the first element in the tuple not being an integer.

File: src/amoc_coordinator_worker.erl (hunk 6 above)
Comment (DenysGonchar) [reply src/amoc_coordinator_worker.erl]: it should be started at the Scenario:init(), AFAIR that process is shortlived. And we should use it later from Scenario:start() context.

File: src/amoc_coordinator.erl (hunk 7 above)
[the remark this answers is not in the document]
Comment (michalwski) [reply src/amoc_coordinator.erl]: Right, I didn't notice the guard. Thanks!

**Comment by DenysGonchar** [comment]
> Oh, one more thing. Is this meant to be used in single AMOC node or in a cluster?

in the multi node environment each node can have it's own coordinator, so it will coordinate only the users of the exact slave amoc node

File: src/amoc_coordinator_worker.erl (hunk 6 above)
Comment (michalwski) [reply src/amoc_coordinator_worker.erl]: And do we care if we loose the `amoc_coordinator_worker` process while the test is running?

Comment (DenysGonchar) [reply src/amoc_coordinator_worker.erl]: sorry, i got i wrong. yes, this one can and should be changed to `start_link`. it's important to have `gen_event` to be just started, but not linked.

**Comment by DenysGonchar** [comment]
> I think this functionality has some potential, it also has a risk of being not used especially when there is no documentation. I'd like to see some documentation with examples showing benefits of this feature.

i was thinking about creation of some simple example scenario, also it will be used in the (at least) 3 available scenarios. however some text description is good idea.  also it would be nice to describe amoc_throttle as well. so i would prefer to create a separate PR with documentation.

also where would you like me to place such documentation? extend Readme.md? repo's wiki pages? create some docs dir?

**Comment by michalwski** [comment]
> also @michalwski where would you like me to place such documentation? extend Readme.md? repo's wiki pages? create some docs dir?

Right, that's good question. So far we have only Readme.md I think the documentation could be put there for now. Later, we probably need to restructure the doc a move some content to dedicated pages in the `doc` dir.

**Pull request merged**